blog: why AI needs sync post. (#2511)
electric-sql/electric: website/blog/posts/2025-04-09-building-ai-apps-on-sync.md

  </video>
</figure>

If, instead, you stream tokens into a store and then subscribe to that store, you can build non-fragile, resilient apps where the data isn't lost when a connection drops.

<figure>
  <img class="hidden-sm"
      src="/img/blog/building-ai-apps-on-sync/streaming-via-store.png"
  />
  <img class="block-sm" style="width: 100%; max-width: 396px"
      src="/img/blog/building-ai-apps-on-sync/streaming-via-store.sm.png"
  />
</figure>

For example, here's our [Electric AI chat app](https://github.com/electric-sql/electric-ai-chat), streaming tokens via a store. It handles offline, patchy connectivity and page refreshes without a problem:

<figure>
  <video controls preload="auto">
    <source src="https://electric-sql-blog-assets.s3.us-east-1.amazonaws.com/building-collaborative-ai-apps-on-sync/electric-chat-resilient-1080.mp4" />
  </video>
</figure>

The key to this behaviour is *resumeability*: the ability to resume streaming from a known position in the stream. To do this, the app keeps track of the last position its seen. Then when re-connecting, it requests the stream from that position.

> This pattern is fiddly to wire up yourself (message delivery is a [distributed systems rabbit hole](https://jepsen.io/consistency/models)) but is *built in* to sync engines for you. For example, Electric's [sync protocol](/docs/api/http) is based on the client sending `offset` and `cursor` parameters.

Reviewer: ```suggestion
This pattern is fiddly to wire up yourself (message delivery is a [distributed systems rabbit hole](https://jepsen.io/consistency/models)) but is *built in* to sync engines for you. For example, Electric's [sync protocol](/docs/api/http) is based on the client sending an `offset` parameter.
```

Cursor is just used for cache busting so I'd remove it. It's a bit confusing too to mention both as people typically either use offsets or cursors to traverse through a result set so people would wonder why we have both.

```tsx
function TodoListItems() {
  const { data: todoListItems } = useShape({
    url: `${ELECTRIC_API_URL}/v1/shape`,
    params: {
      table: 'todo_lists_items'
    }
  })

  return (
    <ul>
      {todoListItems.map((todoListItem) => (
        <li key={todoListItem.id}>
          {todoListItem.task}
          {todoListItem.done && <span> Done</span>}
        </li>
      ))}
    </ul>
  )
}
```

### Structure

So far, when discussing streaming, we've focused on tokens. But models are just as adept at returning structured data. This is another major advantage of streaming through a store. That store can be a structured database.

This allows agents to collaborate on different parts of a shared state, by working on different parts of a structured data model. For example, one agent can be outlining the high level structure of a Figma project whilst another agent fills in the details on each of the canvases.

### Chaos

When you call an API or function you typically know the "blast radius" of what data it can change. So you can know what to refetch. When you interact with an AI agent (that has any kind of agency) you don't know what it's going to change.

> So you either need to constantly track and re-fetch everything. Or you need to monitor what data changes, so that you're automatically informed about it. What you really need is a way of declaring the subset of the data that the app, agent or UI depends on so you can automatically monitor it, stay up-to-date and be able to respond to changes.

Reviewer: ```suggestion
So you either need to constantly track and re-fetch everything. Or you need to monitor what data changes, so that you're automatically informed about it. What you really need is a way of declaring the subset of the data that the app, agent or UI might care about on so you can automatically monitor it, stay up-to-date and be able to respond to changes.
```

That's why Sunil Pai says that [AI agents are local-first clients](https://sunilpai.dev/posts/local-first-ai-agents/) and that's why Theo Brown is [searching for the ideal sync engine](https://youtu.be/3gVBjTMS8FE).

## Sync is the solution

Sync solves a range of practical challenges with AI UX. From resumeability and interruptibility to multi-tab, multi-device and multi-user.

As AI agents become more collaborative and autonomous (and lots more of them are spawned), then sharing state, reviewing progress, reacting to changes and maintaining local data sets are all going to get more important.

> ### To replacing SaaS

Reviewer: The premise here is a bit odd — there's new and old companies building tools. Also AI or not, it's still SaaS. It's perhaps agentic-saas but it's still software doing a job.

Author: The “replace SaaS and enterprise software” is the opportunity for AI apps. Technically AI apps may still be delivered as a service, but SaaS in this sense is more an era / generation of software.

Reviewer: Ok sure — just the rest of the article hasn't been that focused on new AI startups seeking to overturn the old order. I'm not sure how common is for people think specifically in these terms of "replacing saas". It's not something I've heard. 

Given there's tons of AI product teams in existing companies, I'd keep things more general along the lines of "sync is necessary for AI to become the dominant paradigm".

Author: Yup, I've re-written with less "SaaS killer" and more "team-based collaboration is an important feature to unlock b2b and enterprise adoption":

```
### Unlocking adoption

One of the main opportunities for AI startups and teams building AI apps is to replace current-generation software with smarter, AI-driven systems. Particularly in b2b and enterprise software, which tends to be build around team-based collaboration and support for multiple users with different roles.

This is where the ability to build multi-user, collaborative AI apps is key to adoption. Single-user AI sessions are not going to cut it. To replace incumbent systems and get wide adoption across the enterprise, AI apps and agentic systems need to support team-based collaboration. As we've seen, that means keeping multiple users and agents in sync.

Sync is a hard problem to solve yourself &mdash; and the last thing you want to be spending time on when you could be building your core product. That's why AI apps with ambition should be built on a sync engine, like [Electric](/), that solves sync for you.
```

> For AI apps to deliver on their promise of replacing current-generation SaaS and enterprise software, they need to match and then exceed the capabilities of the current software. SaaS has had decades to build out multi-user support and team-based collaboration. Single-user AI sessions are not going to scale across the enterprise.

Reviewer: ```suggestion
For AI apps to deliver on their promise of replacing current-generation SaaS and enterprise software, they need to match and then exceed the capabilities of the current software. SaaS has had decades to build out multi-user support and team-based collaboration. Single-user AI sessions hamstrung the UX.
```

Author: If your GTM is to sell to enterprise, not being able to scale across the enterprise is a pretty sharp motivator. Enterprise software is team oriented, requires users with roles, etc.

Reviewer: Sure yeah, enterprise GTM is good. What is the biggest benefit sync has for someone looking to build an enterprise product? Maybe say something like "Single-user AI is too limited to build sophisticated team oriented enterprise software"
